File: spring-boot-async-and-mockito-spies-java/src/main/java/x/scratch/ScratchApplication.java
package x.scratch;

import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import static x.scratch.AsyncConfiguration.SLOW_EXECUTOR_BEAN_NAME;

@SpringBootApplication
public class ScratchApplication {
    public static void main(final String[] args) {
        SpringApplication.run(ScratchApplication.class, args);
    }

    @Component
    @RequiredArgsConstructor(onConstructor = @__(@Autowired))
    public static class Sally {
        private final Bob bob;

        public void runItEventuallyButQuickly() {
            bob.runItEventuallyButQuickly();
        }

        public void runItEventuallyButSlowly() {
            bob.runItEventuallyButSlowly();
        }
    }

    @Component
    public static class Bob {
        @Async
        public void runItEventuallyButQuickly() {}

        @Async(SLOW_EXECUTOR_BEAN_NAME)
        public void runItEventuallyButSlowly() {}
    }
}

File: spring-boot-async-and-mockito-spies/src/main/kotlin/x/scratch/ScratchApplication.kt
package x.scratch

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component

@SpringBootApplication
class ScratchApplication

fun main(args: Array<String>) {
    runApplication<ScratchApplication>(*args)
}

@Component
class Sally @Autowired constructor(
        private val bob: Bob) {
    fun runItEventuallyButQuickly() {
        bob.runItEventuallyButQuickly()
    }

    fun runItEventuallyButSlowly() {
        bob.runItEventuallyButSlowly()
    }
}

@Component
open class Bob {
    @Async
    open fun runItEventuallyButQuickly() {
    }

    @Async(SLOW_EXECUTOR_BEAN_NAME)
    open fun runItEventuallyButSlowly() {
    }
}
